### DISCUSSION TITLE
General Performance in Insert/Update operations

### DISCUSSION THREAD

**Discussion by parsa-ra** [milvus-io/milvus#10951] [Q&A and General discussion]
Most of the benchmarks you can find on the web usually deal with query time scenarios; I wonder if there is any discussion or any approximate performance of the various indexes in their insertion/Update phase settings.

For example, suppose we have a 1B entry dataset, and on average, we receive 50 update/insertion operations per second. What will be the operating point of re-indexing the whole dataset and linear search on newly inserted data as it grows.

**Comment by xiaofan-luan** [comment]
For Milvus we don't reindex all the dataset since we split original data into segment. Only the incremental data is indexed. 
However it is a interesting test to run indextime consumption. One way to measure the performance is to create a collection without index, add index to the collection and see when the index will be done.

**Reply by parsa-ra** [reply]
I am bringing up the issue by an example, suppose we indexed 10 million vectors using an IVF_SQ8 index, with parameter nlist=16k, now suppose during a period of 1 day we incrementally added another million or so, if we rely on the previous index (16k cluster centers which calculated using 10 million vectors) we definitely losing a bit of accuracy/recall vs re-indexing the whole dataset (11 million) isn't that right (since there might have been an improvement in recall/accuracy if we included the cluster center generation phase with the whole dataset )? My main question here is how we can deal with such incremental data, is there any articles/papers to dive a little bit deeper into this issue? Should we develop an operating point for each dataset based on their distribution?

Please note that the raised issue can be expanded to other indexes like HNSW or ANNOY too.

**Reply by yhmo** [reply]
@parsa-ra 
The milvus splits data into segments, and there is a threshold(default 500MB) to define when a segment can build index. In a buffer there is a "growing" segment, when you incrementally insert vectors,  the segment become larger and larger. When its size reach the threshold, milvus will start index task for the segment asynchronously. Before index task is finished, milvus will performs brute-force search for this segment.
Assume you have insert 10 million vectors, there are already 10 segments in milvus, each segment has 1 million vectors, all the 10 segments have finished index. Then you increment insert another million vectors, the latest segment has no index. Now the client send search request(topk=100), the milvus server will do ANN search in each segment, 10 segments use index, 1 segment use brute-force search. Each segment generate a result(topk=100), there are 11 results. Milvus reduces them, combines them into a final result(topk=100) return to client. Once the growing segment size reach threshold, milvus will build index for it and use its index to search.
Yes, combine results from different index will affect recall/accuracy, this is a workaround to handle incremental data and trade-off between performance and usability.

**Reply by parsa-ra** [reply]
Got it, Thanks for the detailed explanation.

**Reply by xiaofan-luan** [reply]
Just want to give more details, 
Combine segment results won't affect recall, however segment size do affect query performance. (In the extreme case, each segments only has K results, then is falls back to FLAT search).
The tradeoff here is the larger each segments, the harder you can do load balance and slower to build index on top of it.  Another mechanism named compaction is introduced to solve the problem.

**Reply by shauryagoel** [reply]
@xiaofan-luan can you please describe how the indexing works in the above situation?  That there are 10 segments, each with 1 million vectors and each is indexed with `nlist`=16k. Does each of the 10 segments have nlist=16k, or does the whole 10 segments have a single `nlist` as 16k? Similarly, how is `nprobe` used? Also, what happens when the 11th segment needs to be indexed?

**Reply by yhmo** [reply]
> @xiaofan-luan can you please describe how the indexing works in the above situation? That there are 10 segments, each with 1 million vectors and each is indexed with `nlist`=16k. Does each of the 10 segments have nlist=16k, or does the whole 10 segments have a single `nlist` as 16k? Similarly, how is `nprobe` used? Also, what happens when the 11th segment needs to be indexed?

- In milvus, each segment has an independent index. So, per your question, each segment is indexed with 'nlist=16k'
- The client call search(nprobe=100), the search engine will search the segments one by one. Search index of segment_1, find out nprobe nearest buckets, then compute distance to get topk items; Search index of segment_2, find out nprobe nearest buckets, compute distance to get topk items..... Search index of segment_N, find out nprobe nearest buckets, compute distance to get topk items, etc. Finally, merge the n * topk items into a final topk items and return to the client.
- Once the 11th segment is generated, an index node will begin to build index for the segment, with nlist=16k. You still can perform search to the collection. The search engine search the 10 segments one by one by index, and search the 11th segment by brute-force search if its index is not ready.

**Reply by shauryagoel** [reply]
Understood, thanks for your detailed explanation.
